gemini-vertex-model-provider/server/server.go

```diff
@@ -0,0 +1,632 @@
+package server
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	openai "github.com/gptscript-ai/chat-completion-client"
+	"google.golang.org/genai"
+)
+
+const systemPrompt = `You are a task oriented system.
+Be as brief as possible when answering the user.
+Only give the required answer.
+Do not give your thought process.
+Use functions or tools as needed to complete the tasks given to you.
+You are referred to as a tool.
+Do not call functions or tools unless you need to.
+Ensure you are passing the correct arguments to the functions or tools you call.
+Do not move on to the next task until the current task is completed.
+Do not make up arguments for tools.
+Call functions one at a time to make sure you have the correct inputs.`
+
+type server struct {
+	port   string
+	client *genai.Client
+}
+
+func Run(client *genai.Client, port string) error {
+	mux := http.NewServeMux()
+
+	s := &server{
+		client: client,
+		port:   port,
+	}
+
+	mux.HandleFunc("/{$}", s.healthz)
+	mux.HandleFunc("GET /v1/models", s.listModels)
+	mux.HandleFunc("POST /v1/chat/completions", s.chatCompletions)
+	mux.HandleFunc("POST /v1/embeddings", s.embeddings)
+
+	httpServer := &http.Server{
+		Addr:    "127.0.0.1:" + port,
+		Handler: mux,
+	}
+
+	if err := httpServer.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+		return err
+	}
+
+	return nil
+}
+
+func (s *server) healthz(w http.ResponseWriter, _ *http.Request) {
+	_, _ = w.Write([]byte("http://127.0.0.1:" + s.port))
+}
+
+func (s *server) listModels(w http.ResponseWriter, r *http.Request) {
+	content := map[string]any{
+		"data": []map[string]any{
```

Comment:
Can we add the `usage` metadata to these?

Reply:
Done